gsplat/cuda/csrc/rasterization.cu

```diff
@@ -0,0 +1,1245 @@
+                                    const int n_tiles, const int tile_n_bits,
+                                    int32_t *__restrict__ offsets // [C, n_tiles]
+) {
+    // e.g., ids: [1, 1, 1, 3, 3], n_tiles = 6
+    // counts: [0, 3, 0, 2, 0, 0]
+    // cumsum: [0, 3, 3, 5, 5, 5]
+    // offsets: [0, 0, 3, 3, 5, 5]
+    unsigned idx = cg::this_grid().thread_rank();
+    if (idx >= n_isects)
+        return;
+
+    int64_t isect_id_curr = isect_ids[idx] >> 32;
+    int64_t cid_curr = isect_id_curr >> tile_n_bits;
+    int64_t tid_curr = isect_id_curr & ((1 << tile_n_bits) - 1);
+    int64_t id_curr = cid_curr * n_tiles + tid_curr;
+
+    if (idx == 0) {
+        // write out the offsets until the first valid tile (inclusive)
+        for (int i = 0; i < id_curr + 1; ++i)
+            offsets[i] = idx;
+    }
+    if (idx == n_isects - 1) {
+        // write out the rest of the offsets
+        for (int i = id_curr + 1; i < C * n_tiles; ++i)
+            offsets[i] = n_isects;
+    }
+
+    if (idx > 0) {
+        // visit the current and previous isect_id and check if the (cid, tile_id)
+        // pair changes.
+        int64_t isect_id_prev = isect_ids[idx - 1] >> 32; // shift out the depth
+        if (isect_id_prev == isect_id_curr)
+            return;
+
+        // write out the offsets between the previous and current tiles
+        int64_t cid_prev = isect_id_prev >> tile_n_bits;
+        int64_t tid_prev = isect_id_prev & ((1 << tile_n_bits) - 1);
+        int64_t id_prev = cid_prev * n_tiles + tid_prev;
+        for (int i = id_prev + 1; i < id_curr + 1; ++i)
+            offsets[i] = idx;
+    }
+}
+
+torch::Tensor isect_offset_encode_tensor(const torch::Tensor &isect_ids, // [n_isects]
+                                         const int C, const int tile_width,
+                                         const int tile_height) {
+    DEVICE_GUARD(isect_ids);
+    CHECK_INPUT(isect_ids);
+
+    int n_isects = isect_ids.size(0);
+    torch::Tensor offsets = torch::empty({C, tile_height, tile_width},
+                                         isect_ids.options().dtype(torch::kInt32));
+    if (n_isects) {
+        int n_tiles = tile_width * tile_height;
+        int tile_n_bits = (int)floor(log2(n_tiles)) + 1;
+        at::cuda::CUDAStream stream = at::cuda::getCurrentCUDAStream();
+        isect_offset_encode<<<(n_isects + N_THREADS - 1) / N_THREADS, N_THREADS, 0,
+                              stream>>>(n_isects, isect_ids.data_ptr<int64_t>(), C,
+                                        n_tiles, tile_n_bits,
+                                        offsets.data_ptr<int32_t>());
+    } else {
+        offsets.fill_(0);
+    }
+    return offsets;
+}
+
+/****************************************************************************
+ * Rasterization
+ ****************************************************************************/
+
+__global__ void rasterize_to_indices_iter_kernel(
```

Comment:
okay so this function only rasterizes a range of indices determined by step0 to step1
maybe rename to rasterize_to_indices_in_range_kernel and rename step0 and step1 to range_start and range_end respectively?

Reply:
good point. done!